sanitizers/src/test/java/com/example/UnsafeArrayOutOfBounds.java

```diff
@@ -0,0 +1,552 @@
+    }
+
+    static void putShort() {
+      UNSAFE.putShort(new short[5], Unsafe.ARRAY_SHORT_BASE_OFFSET - 1, (short) 0);
+    }
+
+    static void putShort_end() {
+      UNSAFE.putShort(
+          new short[5],
+          Unsafe.ARRAY_SHORT_BASE_OFFSET + 5L * Unsafe.ARRAY_SHORT_INDEX_SCALE,
+          (short) 0);
+    }
+
+    static void putShortVolatile() {
+      UNSAFE.putShortVolatile(new short[5], Unsafe.ARRAY_SHORT_BASE_OFFSET - 1, (short) 0);
+    }
+
+    static void putShortVolatile_end() {
+      UNSAFE.putShortVolatile(
+          new short[5],
+          Unsafe.ARRAY_SHORT_BASE_OFFSET + 5L * Unsafe.ARRAY_SHORT_INDEX_SCALE,
+          (short) 0);
+    }
+
+    static void copyMemory() {
+      UNSAFE.copyMemory(
+          new byte[5],
+          Unsafe.ARRAY_BYTE_BASE_OFFSET - 1,
+          new byte[5],
+          Unsafe.ARRAY_BYTE_BASE_OFFSET,
+          2);
+    }
+
+    static void copyMemory_end() {
+      UNSAFE.copyMemory(
+          new byte[5],
+          Unsafe.ARRAY_BYTE_BASE_OFFSET + 4L * Unsafe.ARRAY_BYTE_INDEX_SCALE,
+          new byte[5],
+          Unsafe.ARRAY_BYTE_BASE_OFFSET,
+          2);
+    }
+
+    static void copyMemory_dest() {
+      UNSAFE.copyMemory(
+          new byte[5],
+          Unsafe.ARRAY_BYTE_BASE_OFFSET,
+          new byte[5],
+          Unsafe.ARRAY_BYTE_BASE_OFFSET - 1,
+          2);
+    }
+
+    static void copyMemory_dest_end() {
+      UNSAFE.copyMemory(
+          new byte[5],
+          Unsafe.ARRAY_BYTE_BASE_OFFSET,
+          new byte[5],
+          Unsafe.ARRAY_BYTE_BASE_OFFSET + 4L * Unsafe.ARRAY_BYTE_INDEX_SCALE,
+          2);
+    }
+
+    static void setMemory() {
+      UNSAFE.setMemory(new byte[5], Unsafe.ARRAY_BYTE_BASE_OFFSET - 1, 2, (byte) 0);
+    }
+
+    static void setMemory_end() {
+      UNSAFE.setMemory(
+          new byte[5],
+          Unsafe.ARRAY_BYTE_BASE_OFFSET + 4L * Unsafe.ARRAY_BYTE_INDEX_SCALE,
+          2,
+          (byte) 0);
+    }
+
+    // The following covers some additional special cases of invalid memory access
+    static void byteAccessOnObjectArray() {
+      UNSAFE.getByte(new String[10], Unsafe.ARRAY_OBJECT_BASE_OFFSET);
+    }
+
+    static void objectAccessOnPrimitiveArray() {
+      UNSAFE.getObject(new byte[100], Unsafe.ARRAY_BYTE_BASE_OFFSET);
+    }
+
+    static void unalignedObjectAccess() {
+      assert Unsafe.ARRAY_OBJECT_INDEX_SCALE != 1;
+      UNSAFE.getObject(new String[2], Unsafe.ARRAY_OBJECT_BASE_OFFSET + 1);
+    }
+  }
+
+  public static void fuzzerTestOneInput(FuzzedDataProvider data) throws Exception {
+    Method[] testMethods = TestMethods.class.getDeclaredMethods();
+    // Since all of these methods are expected to cause a sanitizer exception, pick a random one and
+    // run it
+    // TODO: Is this a proper way to implement this?
+    Method testMethod = testMethods[data.consumeInt(0, testMethods.length - 1)];

```

Comment:
Not sure if this is a good / proper way to ensure that all of the methods cause a sanitizer error.

Once a fuzzing test run finds the first expected failing input (in this case here basically _any input_), does it keep reusing that? In that case it would never run any of the other test methods here.

Is there a better way to solve this, other than adding dozens of separate classes each with their own `fuzzerTestOneInput`?

Reply:
Thanks for the tip! I have adjusted it now. The method names in this list are mostly lexicographically ordered, except for a few cases where I put the corresponding `_end` method name directly below the other method name.